src/main/java/com/gtnewhorizons/gtnhgradle/modules/JVMDowngraderModule.java

```diff
@@ -0,0 +1,595 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import xyz.wagyourtail.jvmdg.gradle.JVMDowngraderExtension;
+import xyz.wagyourtail.jvmdg.gradle.task.DowngradeJar;
+import xyz.wagyourtail.jvmdg.gradle.task.ShadeJar;
+import xyz.wagyourtail.jvmdg.gradle.task.files.DowngradeFiles;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/** JVM Downgrader module. Must be applied after {@link ShadowModule} when shadow is used. */
+public class JVMDowngraderModule implements GTNHModule {
+
+    private static final String SHADOW_JAR_TASK = "shadowJar";
+    private static final String DOWNGRADE_JAR_TASK = "downgradeJar";
+    private static final String SHADE_DOWNGRADED_API_TASK = "shadeDowngradedApi";
+    private static final String REOBF_JAR_TASK = "reobfJar";
+
+    @Override
+    public boolean isEnabled(GTNHGradlePlugin.@NotNull GTNHExtension gtnh) {
+        return gtnh.getModernJavaSyntaxMode()
+            .get()
+            .usesJvmDowngrader();
+    }
+
+    @Override
+    public void apply(GTNHGradlePlugin.@NotNull GTNHExtension gtnh, @NotNull Project project) throws Throwable {
+        final int downgradeTarget = gtnh.getDowngradeTargetVersion()
+            .get();
+        final JavaVersion targetVersion = JavaVersion.toVersion(downgradeTarget);
+        final JvmDowngraderStubsProvider stubsProvider = gtnh.getJvmDowngraderStubsProvider()
+            .get();
+        final boolean shadeStubs = stubsProvider.shouldShadeStubs();
+
+        if (gtnh.configuration.jvmDowngraderStubsProvider.isEmpty()) {
+            project.getLogger()
+                .warn(
+                    "jvmDowngraderStubsProvider is not set. Defaulting to 'gtnhlib'. "
+                        + "Set explicitly to 'gtnhlib', 'shade', or 'external' to suppress this warning.");
+        }
+
+        ensureJvmdgSnapshotRepo(project);
+        applyJvmdgPlugin(project, gtnh, targetVersion);
+        configureCompileDependencies(project, gtnh, downgradeTarget, stubsProvider);
+        final DowngradeTasks downgradeTasks = registerDowngradeTasks(project, gtnh, targetVersion);
+        configureTestTask(project, gtnh, downgradeTasks);
+        final TaskProvider<? extends AbstractArchiveTask> publishableDevJar = configureJarTaskChain(
+            gtnh,
+            project,
+            downgradeTarget,
+            targetVersion,
+            shadeStubs);
+        configureRunTasks(gtnh, project, downgradeTasks, publishableDevJar);
+        validateLombokVersion(gtnh, project);
+    }
+
+    private void applyJvmdgPlugin(Project project, GTNHGradlePlugin.GTNHExtension gtnh, JavaVersion targetVersion) {
+        project.getPluginManager()
+            .apply("xyz.wagyourtail.jvmdowngrader");
```

Comment:
Apply the actual plugin class here, not the stringified name - that way we get more compiletime safety